geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/data/AccumuloFeatureReader.scala

```diff
@@ -12,45 +12,119 @@ import java.util.concurrent.atomic.AtomicBoolean
 
 import org.geotools.data.{FeatureReader, Query}
 import org.locationtech.geomesa.accumulo.index.QueryHints.RichHints
+import org.locationtech.geomesa.accumulo.index.QueryPlanner.SFIter
 import org.locationtech.geomesa.accumulo.index._
 import org.locationtech.geomesa.accumulo.stats._
 import org.locationtech.geomesa.filter.filterToString
-import org.locationtech.geomesa.utils.stats.{MethodProfiling, NoOpTimings, TimingsImpl}
+import org.locationtech.geomesa.security.AuditProvider
+import org.locationtech.geomesa.utils.stats.{MethodProfiling, Timings, TimingsImpl}
 import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
 
-class AccumuloFeatureReader(queryPlanner: QueryPlanner, val query: Query, dataStore: AccumuloDataStore)
-    extends FeatureReader[SimpleFeatureType, SimpleFeature] with MethodProfiling {
+trait AccumuloFeatureReader extends FeatureReader[SimpleFeatureType, SimpleFeature] {
 
-  private val start = System.currentTimeMillis()
   private val closed = new AtomicBoolean(false)
+  private lazy val start = System.currentTimeMillis()
 
-  dataStore.queryTimeoutMillis.foreach(timeout => ThreadManagement.register(this, start, timeout))
-
-  private val writeStats = dataStore.isInstanceOf[StatWriter]
-  implicit val timings = if (writeStats) new TimingsImpl else NoOpTimings
+  timeout.foreach(t => ThreadManagement.register(this, start, t))
 
-  private val iter = profile(queryPlanner.runQuery(query), "planning")
+  def query: Query
+  def timeout: Option[Long]
+  def isClosed: Boolean = closed.get()
 
-  override def next() = profile(iter.next(), "next")
+  protected def closeOnce(): Unit
 
-  override def hasNext = profile(iter.hasNext, "hasNext")
+  override def getFeatureType = query.getHints.getReturnSft
 
   override def close() = if (!closed.getAndSet(true)) {
-    iter.close()
-    dataStore.queryTimeoutMillis.foreach(timeout => ThreadManagement.unregister(this, start, timeout))
-    if (writeStats) {
-      val stat = QueryStat(queryPlanner.sft.getTypeName,
-          System.currentTimeMillis(),
-          filterToString(query.getFilter),
-          QueryStatTransform.hintsToString(query.getHints),
-          timings.time("planning"),
-          timings.time("next") + timings.time("hasNext"),
-          timings.occurrences("next").toInt)
-      dataStore.asInstanceOf[StatWriter].writeStat(stat, dataStore.getQueriesTableName(queryPlanner.sft))
+    timeout.foreach(t => ThreadManagement.unregister(this, start, t))
```

Comment:
just wanted to check that there's no chance of RTE thrown here...didn't look like it digging through the ThreadManagement code but if you think it's possible then:

```try{...} finally { closeOnce()}```